server/src/main/java/org/elasticsearch/common/compress/DeflateCompressor.java

```diff
@@ -88,56 +157,75 @@ public StreamInput streamInput(StreamInput in) throws IOException {
-            final AtomicBoolean closed = new AtomicBoolean(false);
-
+        final Releasable releasable;
+        final Inflater inflater;
+        if (threadLocal) {
+            final ReleasableReference<Inflater> current = inflaterForStreamRef.get();
+            if (current.inUse) {
+                // Nested de-compression streams should not happen but we still handle them safely by using a fresh Inflater
+                inflater = new Inflater(true);
+                releasable = inflater::end;
+            } else {
+                inflater = current.get();
+                releasable = current;
+            }
+        } else {
+            inflater = new Inflater(true);
+            releasable = inflater::end;
+        }
+        return new BufferedInputStream(new InflaterInputStream(in, inflater, BUFFER_SIZE) {
+            @Override
             public void close() throws IOException {
                 try {
                     super.close();
                 } finally {
-                    if (closed.compareAndSet(false, true)) {
-                        // important to release native memory
-                        inflater.end();
-                    }
+                    // We are ensured to only call this once since we wrap this stream in a BufferedInputStream that will only close
+                    // its delegate once
+                    releasable.close();
                 }
             }
-        };
+        }, BUFFER_SIZE);
     }
 
     @Override
-    public StreamOutput streamOutput(OutputStream out) throws IOException {
+    public StreamOutput threadLocalStreamOutput(OutputStream out) throws IOException {
         out.write(HEADER);
-        final boolean nowrap = true;
-        final Deflater deflater = new Deflater(LEVEL, nowrap);
+        final ReleasableReference<Deflater> current = deflaterForStreamRef.get();
+        final Releasable releasable;
+        final Deflater deflater;
+        if (current.inUse) {
+            // Nested compression streams should not happen but we still handle them safely by using a fresh Deflater
+            deflater = new Deflater(LEVEL, true);
+            releasable = deflater::end;
+        } else {
+            deflater = current.get();
+            releasable = current;
+        }
         final boolean syncFlush = true;
         DeflaterOutputStream deflaterOutputStream = new DeflaterOutputStream(out, deflater, BUFFER_SIZE, syncFlush);
         OutputStream compressedOut = new BufferedOutputStream(deflaterOutputStream, BUFFER_SIZE);
         return new OutputStreamStreamOutput(compressedOut) {
             final AtomicBoolean closed = new AtomicBoolean(false);
 
             public void close() throws IOException {
-                try {
-                    super.close();
-                } finally {
-                    if (closed.compareAndSet(false, true)) {
+                if (closed.compareAndSet(false, true)) {
```

Comment:
Need this flag still and move the `super.close` into the flag check also due to a Java8 bug: https://bugs.openjdk.java.net/browse/JDK-8054565 that otherwise forces another `.finish` on the underlying stream after the deflater has already been reset (thus adding another `5` bytes to it and breaking things).